sql: fix panic in inverted indexes

m-schneider opened the pull request:
Before this patch, creating and inverted index on a table with rows
containing JSON would cause a panic.

Now the optimization to reduce memory allocations in
EncodeSecondaryIndexes is modified to allow us to reuse slices, but will
also return the amount of the slice used. This will make sure we don't
have duplicates from reusing a longer slice.

Closes #23190

Release note: None

Commit:
sql: fix panic in inverted indexes

Before this patch, creating and inverted index on a table with rows
containing JSON would cause a panic.

Now the optimization to reduce memory allocations in
EncodeSecondaryIndexes is modified to allow us to reuse slices, but will
also return the amount of the slice used. This will make sure we don't
have duplicates from reusing a longer slice.

Closes #23190

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23279)
<!-- Reviewable:end -->

nvanbenschoten commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsqlrun/indexbackfiller.go, line 189 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23279#-L6Yt-8G17F2xXaH-y4s:-L6Yt-8G17F2xXaH-y4t:b-bfebte) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b07f4b469dd1ce726ba52fca1eca0fd3f03f8a6/pkg/sql/distsqlrun/indexbackfiller.go#L189)):*
> ```Go
> 			}
> 			// We're resetting the length of this slice for variable length indexes such as inverted
> 			// indexes which can append entries to the end of the slice.
> ```

Make a note about what can go wrong if we don't do this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23279)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/distsqlrun/indexbackfiller.go:
```diff
@@ -185,6 +185,9 @@ func (ib *indexBackfiller) runChunk(
 			}
+			// We're resetting the length of this slice for variable length indexes such as inverted
+			// indexes which can append entries to the end of the slice.
+			secondaryIndexEntries = secondaryIndexEntries[:len(mutations)]
```
you're reseting these secondary indexes before using them below?

vivekmenezes commented on pkg/sql/distsqlrun/indexbackfiller.go:
```diff
@@ -185,6 +185,9 @@ func (ib *indexBackfiller) runChunk(
 				ib.rowVals, secondaryIndexEntries); err != nil {
```
The comment above EncodeSecondaryIndexes doesnt say that it can generate a different number of index entries as the indexes passed in. This contract has been completely broken. I'd ask that you change the comment to explain that while it's expected to be the same size passed in, it will return a different size slice. I also feel we should panic at the start of EncodeSecondaryIndexes() for len(indexes) == len(secondaryIndexes)

and a comment above the ```if len(entries) > 1 {```

and while you're at it can you change 

```
if len(entries) > 1 {
 +			secondaryIndexEntries = append(secondaryIndexEntries, entries[1:]...)
 +		}
 +		if len(entries) > 0 {
 +			secondaryIndexEntries[i] = entries[0]
```
to

```
secondaryIndexEntries[i] = entries[0]
if len(entries) > 1 {
 			secondaryIndexEntries = append(secondaryIndexEntries, entries[1:]...)
 		}
```

vivekmenezes replied:
Do you think adding unittests to catch the same problem for all the different call sites of this function are valuable?

vivekmenezes commented on pkg/sql/logictest/testdata/logic_test/inverted_index:
```diff
@@ -458,3 +458,17 @@ SELECT * from d where b @> 'null' ORDER BY a;
+                                         ('{"first_name": "Ernie", "status": "Looking for treats", "location" : "Brooklyn"}');
+
+statement ok
+CREATE INVERTED INDEX dogs on users(user_profile);
```
you can run a select here to check that the inverted index is created properly

m-schneider commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/indexbackfiller.go, line 185 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279#-L6Y10YBjPlZBWYSJnUV-r1-185:-L6bQTTYjivnkCD9BMb3:b-pkezp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b07f4b469dd1ce726ba52fca1eca0fd3f03f8a6/pkg/sql/distsqlrun/indexbackfiller.go#L185)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Do you think adding unittests to catch the same problem for all the different call sites of this function are valuable?
</blockquote></details>

There's only one more call site, for this function in rowwriter which doesn't call it in a loop and file also doesn't seem to have any unit tests associated with it. 

---

*[pkg/sql/distsqlrun/indexbackfiller.go, line 189 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279#-L6Yt-8G17F2xXaH-y4s:-L6bQWDypU692LgVYmcq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b07f4b469dd1ce726ba52fca1eca0fd3f03f8a6/pkg/sql/distsqlrun/indexbackfiller.go#L189)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Make a note about what can go wrong if we don't do this.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/indexbackfiller.go, line 190 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279#-L6Y10YBjPlZBWYSJnUV-r1-190:-L6bQoGJUXjLwL37jOkr:bks5xdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b07f4b469dd1ce726ba52fca1eca0fd3f03f8a6/pkg/sql/distsqlrun/indexbackfiller.go#L190)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

you're reseting these secondary indexes before using them below?
</blockquote></details>

Moved it up.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 474 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279#-L6Y10YBjPlZBWYSJnUW-r1-474:-L6bQzlODNxXDtP75CNP:b-f4e21j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b07f4b469dd1ce726ba52fca1eca0fd3f03f8a6/pkg/sql/logictest/testdata/logic_test/inverted_index#L474)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

you can run a select here to check that the inverted index is created properly
</blockquote></details>

Added for both values.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/distsqlrun/indexbackfiller.go, line 185 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279#-L6Y10YBjPlZBWYSJnUV-r1-185:-L6wcnYn4B_uNh-aBfj0:b7eaoki) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b07f4b469dd1ce726ba52fca1eca0fd3f03f8a6/pkg/sql/distsqlrun/indexbackfiller.go#L185)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

There's only one more call site, for this function in rowwriter which doesn't call it in a loop and file also doesn't seem to have any unit tests associated with it. 
</blockquote></details>

I added the changes we talked about including the panic in EncodeSecondaryIndexes.

---

*[pkg/sql/sqlbase/rowwriter.go, line 102 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279#-L6wdFDsAEY24dBozeBR:-L6wdFDsAEY24dBozeBS:b-9ottj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/697a3693a538056033e6cee39cedaa72f01870ef/pkg/sql/sqlbase/rowwriter.go#L102)):*
> ```Go
> 	colIDtoRowIndex map[ColumnID]int, values []tree.Datum,
> ) (secondaryIndexEntries []IndexEntry, err error) {
> 	if rh.indexEntries == nil {
> ```

I changed it to this because after adding a panic, I learned that there are some code paths which already have index entries allocated.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -99,7 +99,7 @@ func (rh *rowHelper) encodeIndexes(
 	colIDtoRowIndex map[ColumnID]int, values []tree.Datum,
 ) (secondaryIndexEntries []IndexEntry, err error) {
-	if len(rh.indexEntries) != len(rh.Indexes) {
+	if rh.indexEntries == nil {
```
Given that we do have expected cases of rh.indexEntries != nil I feel more comfortable with you leaving the code as is ```if len(rh.indexEntries) != len(rh.Indexes) ```so as to never trigger the panic you've added to the code.

m-schneider commented:
*[pkg/sql/sqlbase/rowwriter.go, line 102 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279#-L6wbJ8CiP8oGPUdL00h-r2-102:-L6xB_qJEbDVNKWb6xwf:b-p19d82) ([raw file](https://github.com/cockroachdb/cockroach/blob/697a3693a538056033e6cee39cedaa72f01870ef/pkg/sql/sqlbase/rowwriter.go#L102)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Given that we do have expected cases of rh.indexEntries != nil I feel more comfortable with you leaving the code as is ```if len(rh.indexEntries) != len(rh.Indexes) ```so as to never trigger the panic you've added to the code.
</blockquote></details>

Changed it back to what it was.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23279)*
<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (approved):
LGTM

jordanlewis commented:
This needs to be cherry-picked into 2.0.
